### ISSUE TITLE
Downloads: Track them, and set a cookie

### ISSUE THREAD

**Issue by povrazor** [ludumdare/ludumdare#83]
Number of downloads should be tracked. Use a simple redirect script.

Also, set a cookie (not a cookie, but a flag in local storage). This is to say a download link was clicked. If the flag is not set, pop up a warning.

This is in response to.

```
<CMLSC> Can you max out the amount of ratings you can give to 1 every 5 minutes.
<CMLSC> Because my game has 26 ratings and only 19 downloads. :/
```

**Comment by djdduty** [comment]
A better idea would be to disregard the client all together. When an author enters a link onto their entry allow them to check "Is this a game link?" or similar, then, rather than setting the link on the page directly to that, direct them at an endpoint in your server which will then return a redirect response to the browser to the actual link the author put (Consider it a url shortener if you will).

This would allow you to track what downloads and game links specific users have visited in the actual user session on the server or even in server persistent storage.

To determine whether a vote might be invalid or not, first verify with the session or persistent storage that the user actually visited a game link.  

This also prevents weirdness of players playing a bunch of games and then rating them later or on a different device / different location. Of course even this isn't 100% reliable, so not much action can be taken beyond warning them and collecting statistics from events like this (which is easy to do if you move it server side) to try and find a solution.

**Comment by povrazor** [comment]
You might have missed it but I actually said both (i.e. the "also"). The default behaviour would be to pop up a warning if you try to rate a game you didn't download. The "cookie" is just the flag to make that warning go away. It would intercept the click, before the redirect.

**Comment by djdduty** [comment]
Oh, my bad then, I didn't see anything about redirect or using the server when I read through it originally. In that case, yes. 

I approve of this message!

**Comment by povrazor** [comment]
To be fair, I was vague.

> Use a simple redirect script.

**Comment by JanWerder** [comment]
I don't really get what the cookie is for? Is the game download logged serverside or not?

**Comment by povrazor** [comment]
Yes, but it's not high priority enough information to do a whole database query over (have I downloaded this yes/no). It's not foolproof (multiple devices, clearing cookies), but the point is one of many little things to discourage rating a game without playing it.

**Comment by JanWerder** [comment]
Okay, so first check if the cookie is present and if it is, do the query?
Mike Kasprzak <notifications@github.com> schrieb am Fr., 12. Aug. 2016 um
14:12:

**Comment by povrazor** [comment]
No, if cookie is present then we know they've downloaded it. No need to pop up the pre-warning "did you download the game?" before accepting votes.

**Comment by JanWerder** [comment]
And if the cookie is not present we do the query?
Mike Kasprzak <notifications@github.com> schrieb am Fr., 12. Aug. 2016 um
14:22:

**Comment by povrazor** [comment]
If cookie is not present, we assume they haven't downloaded it, and give a popup when they click on a rating. No query. 

The whole thing about the cookie is an optimisation and assumption. We actually have no idea if you downloaded it, even if we track it. You could have gone direct, to sites like Itch.io or NewGrounds, or clicked it via a Tweet or Facebook post. The point is that by doing the popup every time we don't know you downloaded, we slowly begin to train users to prefer downloading via the user page, so we can log it.

**Comment by povrazor** [comment]
Just noting this needs to be revisited. Maybe on LD weekend.

**Comment by povrazor** [comment]
Actually naw, I'm going to defer this. But I'll still keep this on my list.

**Comment by povrazor** [comment]
Had an interesting thought on this: Directly disable voting for a game until 15-30 seconds have passed since you clicked a download link.

The point is to be annoying enough to  anyone trying to abuse the system by rapid voting, without being frustrating to users that legitimately don't have a tracked download.
